File: go/day_10/main.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"

	"github.com/maneac/aoc2021/utils/lib/go/bench"
)

const (
	part1Solution = "266301"
	part2Solution = "3404870164"
)

type Input []string

func main() {
	bench.Config{
		Filename:      "./bench/results/go/day_10.csv",
		DataDirectory: "./data",
		ReadData:      func(p string) bench.Day { return readData(p) },
		Part1Solution: part1Solution,
		Part2Solution: part2Solution,
	}.Run()
}

func readData(dir string) Input {
	contents, err := os.ReadFile(filepath.Join(dir, "day_10.txt"))
	if err != nil {
		panic(err)
	}

	return parseContents(string(contents))
}

func parseContents(contents string) Input {
	return strings.Split(contents, "\n")
}

func (i Input) Part1() string {
	score := 0
	for _, line := range i {
		_, illegal := fixLine(line)
		if illegal == 0 {
			continue
		}
		switch illegal {
		case ')':
			score += 3
		case ']':
			score += 57
		case '}':
			score += 1197
		case '>':
			score += 25137
		}
	}
	return fmt.Sprint(score)
}

func (i Input) Part2() string {
	fixScores := make([]int, 0, len(i))
	for _, line := range i {
		fix, illegal := fixLine(line)
		if illegal != 0 {
			continue
		}

		fixScore := 0
		for _, chr := range fix {
			fixScore *= 5
			switch chr {
			case ')':
				fixScore += 1
			case ']':
				fixScore += 2
			case '}':
				fixScore += 3
			case '>':
				fixScore += 4
			}
		}

		fixScores = append(fixScores, fixScore)
	}

	sort.Ints(fixScores)
	return fmt.Sprint(fixScores[len(fixScores)/2])
}

func fixLine(line string) (string, byte) {
	missing := ""
	for len(line) > 0 {
		chr := line[0]
		line = line[1:]

		switch chr {
		case '(':
			missing = ")" + missing
		case '[':
			missing = "]" + missing
		case '{':
			missing = "}" + missing
		case '<':
			missing = ">" + missing
		case ')', ']', '}', '>':
			match := missing[0]
			missing = missing[1:]
			if match != chr {
				return "", chr
			}
		}
	}
	return missing, 0
}

File: ts/day_10/main.ts
function readData(): string[] {
  const contents = Deno.readTextFileSync("./data/day_10.txt").trim();

  return parseContents(contents);
}

function parseContents(contents: string): string[] {
  return contents.split("\n");
}

function part1(data: string[]): number {
  return data.map((line): number => {
    try {
      fixLine(line);
    } catch (e) {
      switch (e) {
        case ")":
          return 3;
        case "]":
          return 57;
        case "}":
          return 1197;
        case ">":
          return 25137;
      }
    }
    return 0;
  }).reduce((acc, cur) => acc + cur);
}

function part2(data: string[]): number {
  const fixScores = data.map((line) => {
    try {
      const fix = fixLine(line);
      return fix.split("").reduce((acc, cur) => {
        switch (cur) {
          case ")":
            return (acc * 5) + 1;
          case "]":
            return (acc * 5) + 2;
          case "}":
            return (acc * 5) + 3;
          case ">":
            return (acc * 5) + 4;
          default:
            throw new Error("invalid character in fix: " + cur);
        }
      }, 0);
    } catch {
      return 0;
    }
  }).filter((fixScore) => fixScore > 0);

  return fixScores.sort((a, b) => a - b)[(fixScores.length - 1) / 2];
}

function fixLine(line: string): string {
  let missing = "";
  while (line.length > 0) {
    const chr = line[0];
    line = line.slice(1);

    switch (chr) {
      case "(":
        missing = ")" + missing;
        break;
      case "[":
        missing = "]" + missing;
        break;
      case "{":
        missing = "}" + missing;
        break;
      case "<":
        missing = ">" + missing;
        break;
      case ")":
      case "]":
      case "}":
      case ">": {
        const match = missing[0];
        missing = missing.slice(1);
        if (match != chr) {
          throw chr;
        }
        break;
      }
    }
  }
  return missing;
}

function main() {
  const data = readData();

  console.log("Part 1: ", part1(data));
  console.log("Part 2: ", part2(data));
}

export { main, parseContents, part1, part2, readData };
